clients/src/main/java/org/apache/kafka/common/utils/CollectionUtils.java

```diff
@@ -67,13 +66,29 @@ private CollectionUtils() {}
         Map<String, List<Integer>> partitionsByTopic = new HashMap<>();
         for (TopicPartition tp: partitions) {
             String topic = tp.topic();
-            List<Integer> topicData = partitionsByTopic.get(topic);
-            if (topicData == null) {
-                topicData = new ArrayList<>();
-                partitionsByTopic.put(topic, topicData);
-            }
+            List<Integer> topicData = partitionsByTopic.computeIfAbsent(topic, k -> new ArrayList<>());
             topicData.add(tp.partition());
         }
         return  partitionsByTopic;
     }
+
+    /**
+     * Abstract duplicate code from StickAssignor and ConsumerProtocol
+     *
+     * @param struct                 data source
+     * @param topicPartitionsKeyName Topic Partitions key names
+     * @return list of TopicPartition from struct
+     */
+    public static List<TopicPartition> getTopicPartitionFromStruct(Struct struct, String topicPartitionsKeyName) {
```

Comment:
I'd rather put this into `org.apache.kafka.common.requests.RequestUtils` given the parameters of your method header.

Reply:
RequestUtils is a package-private class, it can't be used in other package like `org.apache.kafka.clients`

Reply:
I agree with @viktorsomogyi that `RequestUtils` would be a better location. Maybe we can just make it public? The package is internal anyway.

By the way, the code duplication is definitely a little annoying, but utility functions are not a great fit for the current schema APIs because it breaks the encapsulation of the schema. For example, now the key names have to be split across multiple classes. And it's a little odd that only the key for the array is passed. I would probably just leave the duplication in order to preserve the schema encapsulation.